### PR TITLE
Extract categorical feature identification & add cat_features to catboost

### PR DESCRIPTION
**Pull request by RMHogervorst**
Hi there! I have a small proposal and I hope you consider it.  Greetings, Roel

In this PR 
Commit1 : I extract the categorical column identification function, split it into two separate functions: one for detection and one for replacement of factors with integers. 
I think (and I leave it for you to decide if that is true) it is better to have the two things pulled apart. It is easier to test the functionality and we reuse the functions. Also because catboost apperantly uses 0-indexed columns, the categorical indices are easily modified for catboost vs lightgbm while the underlying logic remains unduplicated. 

commit2: Add categorical features to catboost and replace the categorical functionality currently present in lightgbm. I also added the same type of prepare_df functionality in catboost like lightgm so the two are more comparable.

### CODE REVIEW

**Comment by RMHogervorst** [comment]
Related #7

**Comment by RMHogervorst** [comment]
wow. I completely missed that! I don't think we should override that behavior, 
however for cat_features arg: 

> NULL (it is assumed that all columns are the values of numerical features)

unfortunately I cannot inspect the resulting dataframe of load_pool

**Comment by RMHogervorst** [comment]
and in the examples they do turn categorical columns into numeric and pass the column id.

**Comment by RMHogervorst** [comment]
I did a quick check and I think you do not need to supply the categorical feature in training. 
I wrote the exploration [here](https://blog.rmhogervorst.nl/blog/2020/07/22/how-does-catboost-deal-with-factors/).

**Comment by dfalbel** [comment]
Thanks @RMHogervorst your exploration is very revealing. Awesome! It's great that they are equivalent.
So we have 2 options, using the builtin factor to integer convertion or implement our own (as you did in this PR)

I don't have any strong opinions over this. Pros for implementing ourr own is that we know exactly what happens and don't depend on them changing something. Cons: we have maintain this code.

What do you think?

**Comment by RMHogervorst** [comment]
My intuition would be to leave the conversion to catboost, but do supply the categorical features indices. (that way you can override the behavior, and categorize numeric columns if you want to. (I'll modify the PR)

**Comment by dfalbel** [comment]
Agreed! That makes sense to me.

File: R/catboost.R [hunk 1]
```diff
@@ -194,6 +194,27 @@ add_boost_tree_catboost <- function() {
   )
 }
 
+prepare_df_catboost <- function(x, y = NULL, categorical_cols= NULL) {
```
Comment (dfalbel) [inline R/catboost.R:197]: Shoudn't we call prepare_df_catboost in `predict.catboost.Model` too?

Comment (RMHogervorst) [reply R/catboost.R:197]: I am sooo glad there are multiple eyes on this!

Comment (RMHogervorst) [reply R/catboost.R:197]: This calls for more tests 👀

Comment (RMHogervorst) [reply R/catboost.R:197]: I wish the resulting data structure in lightgbm and catboost was more inspectable from R

Comment (dfalbel) [reply R/catboost.R:197]: yeah, it would much easier! It seems to be just a pointer to some C data structure...

**Comment by dfalbel** [comment]
Thanks @RMHogervorst 
Please, can you add yourself as an author in the DESCRIPTION file?

**Pull request merged**